### ISSUE TITLE
Gmic isn't recognized in Gimp 2.9.8

### ISSUE THREAD

**Issue by vivadavid** [dtschump/gmic-community#113]
Hello,

I have unzipped Gmic x64 for Gimp 2.9.8 and it isn't recognized: I tried both Gmic 2.15 and Gmic 2.16 pre. I unzipped the content in:

C:\Users\user_name\AppData\Roaming\GIMP\2.9\plug-ins

I renamed the unzipped folder as <gmic_gimp_qt> to match the name of the EXE file contained in it, and I made sure the folder was listed in Gimp settings (under plug-in folders).

When I open Gimp, I get some some error messages. They are in Spanish and I don't quite understand the terminology, but they refer to objects that can't be found.

Thank you for your help.

Regards,

David
![2017-12-19 - 12 36 57 - gimp startup](https://user-images.githubusercontent.com/9334287/34155460-b26c0554-e4b9-11e7-9be1-d2d60a16c045.png)
![2017-12-19 - 12 36 48 - gimp startup](https://user-images.githubusercontent.com/9334287/34155461-b28b8b36-e4b9-11e7-98da-8d96ec56ad5d.png)
![2017-12-19 - 12 36 36 - gimp startup](https://user-images.githubusercontent.com/9334287/34155462-b2a8ddbc-e4b9-11e7-84cd-3d46853bd7c1.png)
![2017-12-19 - 12 36 23 - gimp startup](https://user-images.githubusercontent.com/9334287/34155463-b3180f0c-e4b9-11e7-8f05-47302f8538fe.png)

**Comment by Jehan** [comment]
Hi!

I don't speak Spanish, but I guess that's about some conflict of libraries. All the functions named there are from glib.
I notice G'Mic duplicates a lot of libraries which are already shipped with GIMP in your zip file for Windows. In particular I even notice a libglib-2.0-0.dll! Why are these duplicated?

**Comment by vivadavid** [comment]
Hello,

I've just tried with Gmic 2.1.7 x64 pre and the result is exactly the same: I get four errores messages and Gmic isn't recgnized.

Regads,

David

**Comment by sylale** [comment]
Bonjour,
I made a new archive with new DLLs.
Informations (french) :
[http://samjcreations.blogspot.ca/2017/12/version-de-test-de-gmic-qt-gimp.html](http://samjcreations.blogspot.ca/2017/12/version-de-test-de-gmic-qt-gimp.html)

Merci.

**Comment by vivadavid** [comment]
Bonjour,

It works!

Merci beaucoup!!

**Comment by vivadavid** [comment]
Hello,

I've tried with the recently releasex 2.1.7 stable, but I get the same error messages. For now, I'll stick to the ZIP provided by sylale (thank you again).

**Comment by sylale** [comment]
Bonjour,

A solution :
1 - Download and install from [http://samjcreations.blogspot.ca/2017/12/version-de-test-de-gmic-qt-gimp.html](http://samjcreations.blogspot.ca/2017/12/version-de-test-de-gmic-qt-gimp.html)

2 - Download and extract [http://gmic.eu/files/windows/gmic_gimp_qt_win64.zip](http://gmic.eu/files/windows/gmic_gimp_qt_win64.zip)

3 - Copy **gmic_gimp_qt.exe** and paste to erase the old executable 

![naj_gmic_qt_gimp](https://user-images.githubusercontent.com/12661733/34279220-8395bc50-e67d-11e7-9032-fe98fdcc138a.png)

Merci.

**Comment by Jehan** [comment]
Please no, do NOT extract the whole content directly under `GIMP/2.9/plug-ins/` as I see in your screenshot. And in particular do not ask people to do this.

I have made a change (in GIMP tree since May so that's available in last development versions) so that it is now possible to have a plug-in inside a subdirectory. All you need to make sure is that the directory is named the same as the executable (minus the ".exe") so that GIMP can find the entry point.
In other words, have a directory `GIMP/2.9/plug-ins/gmic_gimp_qt/` and the plug-in whole content should be inside this directory.

This way, we avoid the DLL hell which can mess up other plug-ins in particular.

**Comment by vivadavid** [comment]
Hi, Sylvie,

I thought I had followed your instructions, and the result was that I got two error messages (instead of the previous four). Then I realized that I had changed the order, I tried again, and it seems everything works now. This is what I did:

1. I unzipped the whole content of your ZIP.
2. I unzipped the EXE file of the official release, which replaced the EXE of your ZIP.

Thank you very much!

One question: will this operation be necessary until a particular future Gmic release?

Regards,

David

P.S. I have placed everything in <GIMP/2.9/plug-ins/gmic_gimp_qt>.

**Comment by sylale** [comment]
Bonjour,

@Jehan
Je sais et j'ai essayé cette modification mais malheureusement cela ne fonctionne pas : les DLL ne sont pas trouvées par Gimp 2.9. La solution proposée ici est provisoire... Le problème est surtout pour les plugins développées en C++. Vous pourriez peut-être proposer un moyen automatique de compiler ces plugins. Personnellement je n'ai pas ces problèmes car mon environnement de compilation est cohérent et je compile tout.
Merci pour votre travail.

@vivadavid 
Erase all the content of the plug-in folder and retry.
Copy **only** gmic_gimp_qt.exe and paste to erase the old executable.
Merci

**Comment by Jehan** [comment]
> Je sais et j'ai essayé cette modification mais malheureusement cela ne fonctionne pas : les DLL ne sont pas trouvées par Gimp 2.9

@sylale Pourtant ça marche ici sous Linux et @vivadavid dit que ça marche aussi pour lui sous Windows dans un sous-répertoire. :-)
Faudrait que j'en sache plus pour comprendre le problème rencontré.

> Vous pourriez peut-être proposer un moyen automatique de compiler ces plugins.

On a un outil (gimptool-2.0), mais effectivement je pense qu'il est pour les plug-ins C. Ceci dit, je pense qu'il peut déjà être utilisé au moins pour avoir les flags de compilation/link avec `gimptool-2.0 --cflags --libs` qui peuvent aussi être utilisés dans une compilation C++.
Mais oui dans l'absolu, pas mal de choses peuvent être améliorées pour la création et gestion de plug-ins GIMP. C'est en projet.

>  Merci pour votre travail.

Et merci du vôtre!

**Comment by sylale** [comment]
Bonjour,
I made a mistake in my previous post in French language.
**The solution to put all elements of the plugin in a single subdirectory works.**
My error was: I only put the DLLs in this single subdirectory that was in the same tree as the executable.
Thank you for insisting.
I will modify my instructions to follow your recommendations.
Merci Jehan et bonne journée.

**Comment by vivadavid** [comment]
Hello, again,

After formatting my computer, I decided to try Gimp 2.10 RC1 and G'mic 2.2.2, but I get the same error messages as before. I unzipped G'mic on <C:\Users\[username]\AppData\Roaming\GIMP\2.10\plug-ins\gmic_gimp_qt> (so that foldername matches de EXE file), and the paths on settings in Gimp are supposed to be OK:

C:\Users\vivad\AppData\Roaming\GIMP\2.10\plug-ins
C:\Program Files\GIMP 2.10\lib\gimp\2.0\plug-ins

Am I doing anything wrong?

Thank you in advance for your time!

**Comment by sylale** [comment]
@vivadavid 

Bonjour,
The version compiled by David is for Gimp 2.8.x and does not work with Gimp 2.10.0 RC1.
In a few hours I will update GimpEval on  [https://samjcreations.blogspot.ca/](https://samjcreations.blogspot.ca/)
Partha may do the same on   [https://www.partha.com/](https://www.partha.com/)

:o)

**Comment by vivadavid** [comment]
Thank you @sylale !

I don't know anything about the internals of G'mic  and Gimp, but shouldn't the official version of G'mic work with Gimp? Last time, I sorted out the problem by using the compilation you offered, but I wonder if this is only because Gimp 2.10 hasn't been officially released.

**Comment by vivadavid** [comment]
Hello,

I've tried Gmic 2.2.2 on Gimp 2.10 and it still doesn't work on this final release. I've tried both the installable and the portable Gmic versions.

I'll wait for the next Gmic release hoping it supports Gimp 2.10.

**Comment by dtschump** [comment]
Thanks for reporting.
The release of GIMP 2.10 means quite a lot of work for us, in order to update the way G'MIC is built, for Linux and Windows. This will probably take a bit of time before the G'MIC binaries provided on the official web pages become compatible with 2.10.
Hopefully, there might be other sources of download sooner (Partha and @sylale pages probably).

**Comment by vivadavid** [comment]
Thank you for your quick reply, @dtschump 

I have no problem waiting, but of course, I'd try other binaries if they are provided.

I'd like to take this opportunity to ask a question: is there a difference between Gmic for Krita and Gmic for Gimp?I read on a Krita changelog that Gmic wasn't fully implemented, but I refer, more specifically, to performance. Applying filters on Gmic (at least, for Krita) takes a long time. My PC is quite old (first generation Intel Core i7) and I understand Gmic filters tend to be quite complex. Should these filters take as long on Gmic for Gimp?

Thanks!

**Comment by dtschump** [comment]
There are only a few difference between the Krita and GIMP plug-ins. The codebase is actually the same, the only thing that differs is the "protocol" of transferring data from the host software to the plug-in (and vice-versa).
Right now, the Krita version does not manage all the data brought by the plug-in, including the positions of the output layers, their opacity and blending mode. So, if you use a G'MIC filter that specifically set a blending mode for one output layer, it won't be considered at all in Krita. This may (hopefully) change in the future, but I guess the Krita devs are more skilled than us to fix those issues.

Concerning the processing time : it should be exactly the same on Krita and GIMP, **if the optimization flags used are the same**. We now compile G'MIC for GIMP with `-Ofast`, so with maximum speed. The version we provide for Krita on the G'MIC website is compiled with `-Ofast` too. I cannot say for other sources providing binaries (I know the Krita team provides an appimage for the G'MIC plug-in for instance).

It could be interesting to have comparison benchmarks, that would say basically if the compilation flags used were the same or not (as the code is exactly the same).
A last note anyway : G'MIC does all its calculations with float-valued versions of the input images, so for big images that represents a big amount of data. G'MIC also uses multi-threaded calculations (through the use of _OpenMP_), which also makes a **huge** difference if you have a lot of cores on your CPU (I have 24 cores on mine, and most filters runs relatively smoothly ;) ).

**Comment by vivadavid** [comment]
Thank you for your thorough answer, @dtschump 

I'd be willing to do some comparison benchmarks once I get Gmic working on Gimp and Krita. In any case, it seems I need a new computer, as certain programmes are too CPU or GPU intensive for me. One of the new i5-i7 8th-generation Intel core processors should give me plenty of power.

As I said, I'm not in a hurry, as I need to get a new computer before I
